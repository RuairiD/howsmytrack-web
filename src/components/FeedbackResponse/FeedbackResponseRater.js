import React, { useState, useMemo } from "react";
import { useMutation } from "react-query";
import axios from "axios";

import { Button, Rate } from "antd";
import { Span } from "lemon-reset";
import apiRoot from "../../apiRoot";

const RATE_FEEDBACK_RESPONSE_MUTATION = `mutation RateFeedbackResponse($feedbackResponseId: Int!, $rating: Int!) {
    rateFeedbackResponse(feedbackResponseId: $feedbackResponseId, rating: $rating) {
        success
        error
    }
}`;

const RATING_TOOLTIP_TEXTS = [
    "This feedback is irrelevant and completely unhelpful.",
    "This feedback is vague, unhelpful, irrelevant or poorly written.",
    "This feedback is useful.",
    "This feedback is relevant, helpful and thoughtful.",
    "This feedback is deeply thoughtful, well-written and very constructive.",
];

const FeedbackResponseRater = ({
    feedbackResponseId,
    currentRating,
}) => {
    const [rating, setRating] = useState(currentRating);

    const onRatingChange = (newRating) => {
        setRating(newRating);
    };

    const submitRating = () => (
        axios.post(`${apiRoot()}/graphql/`, {
            query: RATE_FEEDBACK_RESPONSE_MUTATION,
            variables: {
                feedbackResponseId,
                rating,
            },
        }).then((response) => response.data.data.rateFeedbackResponse)
    );

    const [submitRatingMutate, { data, isLoading }] = useMutation(submitRating);

    const submitted = useMemo(() => (!!currentRating || (data && data.success)), [currentRating, data]);

    return (
        <Span style={{ float: "left", paddingBottom: "0.25em" }}>
            <Rate
                style={{
                    color: "#000000",
                    marginRight: "1em",
                }}
                allowClear
                tooltips={RATING_TOOLTIP_TEXTS}
                value={rating}
                disabled={submitted || isLoading}
                onChange={onRatingChange}
            />
            <Button
                type="primary"
                loading={isLoading}
                disabled={submitted || isLoading || !rating}
                onClick={submitRatingMutate}
            >
                {submitted && "Rated"}
                {!submitted && "Rate"}
            </Button>
        </Span>
    );
};

export default FeedbackResponseRater;
