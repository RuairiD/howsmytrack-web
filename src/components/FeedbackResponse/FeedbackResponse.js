import React, { useState } from "react";

import { Card, Col, Row, Typography } from "antd";
import { Div } from "lemon-reset";

import ViewRepliesButton from "../ViewRepliesButton/ViewRepliesButton";
import FeedbackResponseRepliesModal from "../FeedbackResponseRepliesModal/FeedbackResponseRepliesModal";
import FeedbackResponseRater from "./FeedbackResponseRater";

export type FeedbackResponseProps = {
    feedbackResponseId: number,
    feedback: string,
    currentRating: number,
    allowReplies: boolean,
    replies: number,
    unreadReplies: number,
};

const FeedbackResponse = ({
    feedbackResponseId,
    feedback,
    currentRating,
    allowReplies,
    replies,
    unreadReplies,
}: FeedbackResponseProps) => {
    /*
     * Component for showing user the feedback they have received from another user.
     * Only displayed in a feedback group once the user who received the feedback has
     * left feedback for everyone else in the group.
     */
    const [isRepliesModalVisible, setIsRepliesModalVisible] = useState(false);

    const showRepliesModal = () => {
        setIsRepliesModalVisible(true);
    };

    const onRepliesModalCancel = () => {
        setIsRepliesModalVisible(false);
    };

    return (
        <Card>
            <Row gutter={[16, 16]}>
                <Col>
                    <Typography.Paragraph style={{
                        overflowWrap: "break-word",
                        wordWrap: "break-word",
                    }}
                    >
                        "{feedback}"
                    </Typography.Paragraph>
                </Col>
            </Row>
            <Card.Meta
                title="How helpful was this feedback?"
                description={(
                    <Div style={{ display: "inline" }}>
                        <FeedbackResponseRater
                            feedbackResponseId={feedbackResponseId}
                            currentRating={currentRating}
                        />
                        {(allowReplies && !!currentRating) && (
                            <ViewRepliesButton
                                replies={replies}
                                unreadReplies={unreadReplies}
                                onClick={showRepliesModal}
                            />
                        )}
                    </Div>
                )}
            />
            {allowReplies && (
                <FeedbackResponseRepliesModal
                    feedbackResponseId={feedbackResponseId}
                    feedback={feedback}
                    onCancel={onRepliesModalCancel}
                    isVisible={isRepliesModalVisible}
                />
            )}
        </Card>
    );
};

export default FeedbackResponse;
