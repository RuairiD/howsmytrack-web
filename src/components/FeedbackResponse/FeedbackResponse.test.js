import React from "react";
import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { act } from "react-dom/test-utils";
import FeedbackResponse from "./FeedbackResponse";

describe("FeedbackResponse", () => {
    it("renders a piece of received feedback and a FeedbackResponseRater for that feedback", () => {
        const wrapper = shallow(
            <FeedbackResponse
                feedbackResponseId={1901}
                feedback="feedback"
                currentRating={5}
                allowReplies
                replies={3}
                unreadReplies={2}
            />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("does not render a ViewRepliesButton if replies were not allowed by the original feedback writer", () => {
        const wrapper = shallow(
            <FeedbackResponse
                feedbackResponseId={1901}
                feedback="feedback"
                currentRating={5}
            />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("does not render a ViewRepliesButton if the feedback has not been rated", () => {
        const wrapper = shallow(
            <FeedbackResponse
                feedbackResponseId={1901}
                feedback="feedback"
            />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("shows a FeedbackResponseRepliesModal when ViewRepliesButton is clicked", () => {
        const wrapper = mount(
            <FeedbackResponse
                feedbackResponseId={1901}
                feedback="feedback"
                currentRating={5}
                allowReplies
                replies={3}
                unreadReplies={2}
            />,
        );

        act(() => wrapper.find("ViewRepliesButton").get(0).props.onClick());
        wrapper.update();

        expect(wrapper.find("FeedbackResponseRepliesModal").get(0).props.isVisible).toBeTruthy();
    });

    it("hides the FeedbackResponseRepliesModal when dismissed", () => {
        const wrapper = mount(
            <FeedbackResponse
                feedbackResponseId={1901}
                feedback="feedback"
                currentRating={5}
                allowReplies
                replies={3}
                unreadReplies={2}
            />,
        );

        act(() => wrapper.find("ViewRepliesButton").get(0).props.onClick());
        wrapper.update();
        act(() => wrapper.find("FeedbackResponseRepliesModal").get(0).props.onCancel());
        wrapper.update();

        expect(wrapper.find("FeedbackResponseRepliesModal").get(0).props.isVisible).not.toBeTruthy();
    });
});
