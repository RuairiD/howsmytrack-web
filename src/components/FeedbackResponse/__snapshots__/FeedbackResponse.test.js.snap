// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbackResponse does not render a ViewRepliesButton if replies were not allowed by the original feedback writer 1`] = `
<Card>
  <Row
    gutter={
      Array [
        16,
        16,
      ]
    }
  >
    <Col>
      <Paragraph
        style={
          Object {
            "overflowWrap": "break-word",
            "wordWrap": "break-word",
          }
        }
      >
        "
        feedback
        "
      </Paragraph>
    </Col>
  </Row>
  <Meta
    description={
      <Div
        className=""
        style={
          Object {
            "display": "inline",
          }
        }
      >
        <FeedbackResponseRater
          currentRating={5}
          feedbackResponseId={1901}
        />
      </Div>
    }
    title="How helpful was this feedback?"
  />
</Card>
`;

exports[`FeedbackResponse does not render a ViewRepliesButton if the feedback has not been rated 1`] = `
<Card>
  <Row
    gutter={
      Array [
        16,
        16,
      ]
    }
  >
    <Col>
      <Paragraph
        style={
          Object {
            "overflowWrap": "break-word",
            "wordWrap": "break-word",
          }
        }
      >
        "
        feedback
        "
      </Paragraph>
    </Col>
  </Row>
  <Meta
    description={
      <Div
        className=""
        style={
          Object {
            "display": "inline",
          }
        }
      >
        <FeedbackResponseRater
          feedbackResponseId={1901}
        />
      </Div>
    }
    title="How helpful was this feedback?"
  />
</Card>
`;

exports[`FeedbackResponse renders a piece of received feedback and a FeedbackResponseRater for that feedback 1`] = `
<Card>
  <Row
    gutter={
      Array [
        16,
        16,
      ]
    }
  >
    <Col>
      <Paragraph
        style={
          Object {
            "overflowWrap": "break-word",
            "wordWrap": "break-word",
          }
        }
      >
        "
        feedback
        "
      </Paragraph>
    </Col>
  </Row>
  <Meta
    description={
      <Div
        className=""
        style={
          Object {
            "display": "inline",
          }
        }
      >
        <FeedbackResponseRater
          currentRating={5}
          feedbackResponseId={1901}
        />
        <ViewRepliesButton
          onClick={[Function]}
          replies={3}
          unreadReplies={2}
        />
      </Div>
    }
    title="How helpful was this feedback?"
  />
  <FeedbackResponseRepliesModal
    feedback="feedback"
    feedbackResponseId={1901}
    isVisible={false}
    onCancel={[Function]}
  />
</Card>
`;
